lib/python/qmk/cli/generate/compilation_database.py

```diff
@@ -0,0 +1,125 @@
+    for line in f:
+        if state == 'start':
+            m = file_re.search(line)
+            if m:
+                this_file = m.group(1)
+                state = 'cmd'
+
+        if state == 'cmd':
+            assert this_file
+            m = cmd_re.search(line)
+            if m:
+                # we have a hit!
+                this_cmd = m.group(1)
+                args = shlex.split(this_cmd)
+                args += ['-I%s' % s for s in system_libs(args[0])]
+                new_cmd = ' '.join(shlex.quote(s) for s in args if s != '-mno-thumb-interwork')
+                records.append({"directory": str(QMK_FIRMWARE.resolve()), "command": new_cmd, "file": this_file})
+                state = 'start'
+
+    return records
+
+
+@cli.argument('-kb', '--keyboard', help='The keyboard to build a firmware for. Ignored when a configurator export is supplied.')
+@cli.argument('-km', '--keymap', help='The keymap to build a firmware for. Ignored when a configurator export is supplied.')
+@cli.subcommand('Create a compilation database.')
+@automagic_keyboard
+@automagic_keymap
+def generate_compilation_database(cli):
+    """Creates a compilation database for the given keyboard build.
+
+    Does a make clean, then a make -n for this target and uses the dry-run output to create
+    a compilation database (compile_commands.json). This file can help some IDEs and
+    IDE-like editors work better. For more information about this:
+
+        https://clang.llvm.org/docs/JSONCompilationDatabase.html
+    """
+    command = None
+    # check both config domains: the magic decorator fills in `generate_compilation_database` but the user is
+    # more likely to have set `compile` in their config file.
+    current_keyboard = cli.config.generate_compilation_database.keyboard or cli.config.user.keyboard
+    current_keymap = cli.config.generate_compilation_database.keymap or cli.config.user.keymap
+
+    if current_keyboard and current_keymap:
+        # Generate the make command for a specific keyboard/keymap.
+        command = create_make_command(current_keyboard, current_keymap, dry_run=True)
+
+    elif not current_keyboard:
+        cli.log.error('Could not determine keyboard!')
+    elif not current_keymap:
+        cli.log.error('Could not determine keymap!')
+
+    if command:
+        # remove any environment variable overrides which could trip us up
+        env = os.environ.copy()
+        env.pop("MAKEFLAGS", None)
+
+        # re-use same executable as the main make invocation (might be gmake)
+        clean_command = [command[0], 'clean']
+        cli.log.info('Making clean with {fg_cyan}%s', ' '.join(clean_command))
+        subprocess.run(clean_command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, env=env, check=True)
+
+        cli.log.info('Gathering build instructions from {fg_cyan}%s', ' '.join(command))
+        with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, env=env) as proc:
```

Comment:
I don't have time to do a full review right now, but please use `cli.run` here as it handles running commands on windows correctly. Using subprocess directly in a way that works cross-platform is error-prone.

Reply:
Good call, I wasn't aware of milc's `cli.run`. That'll be straightforward to fix.

As for cross-platform compatibility, I wasn't thinking of that at all. I'll have to spin up a windows environment to test it, but I don't think the way @xton searched for the system library path is very portable. Not sure how I"ll tackle that yet but any suggestions are welcome.